// kueri/dist/src/query-parser/query-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseFilter = exports.parseSort = exports.parseInclude = exports.parsePage = void 0;
function parsePage(page) {
    var _a, _b;
    return {
        size: Number((_a = page === null || page === void 0 ? void 0 : page.size) !== null && _a !== void 0 ? _a : 10),
        number: Number((_b = page === null || page === void 0 ? void 0 : page.number) !== null && _b !== void 0 ? _b : 1),
    };
}
exports.parsePage = parsePage;
function parseInclude(include) {
    return include ? include.split(',') : [];
}
exports.parseInclude = parseInclude;
function parseSort(sort) {
    return sort
        ? Object.fromEntries(sort.split(',').map((prop) => {
            const isDesc = prop.charAt(0) === '-';
            return [isDesc ? prop.slice(1) : prop, isDesc ? 'desc' : 'asc'];
        }))
        : {};
}
exports.parseSort = parseSort;
function parseFilter(filter) {
    return filter !== null && filter !== void 0 ? filter : {};
}
exports.parseFilter = parseFilter;
